enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,407 @@
+part of NTO now ([see here for more info](../node-tuning/pao-in-nto.md)) this
+affords us the chance to consolidate the configuration for `kubelet` and `crio`.
+
+We will modify the code path that generates the [new machine
+config](https://github.com/openshift/cluster-node-tuning-operator/blob/a780dfe07962ad07e4d50c852047ef8cf7b287da/pkg/performanceprofile/controller/performanceprofile/components/machineconfig/machineconfig.go#L91-L127)
+using the performance profile. With the new `workloads.enablePinning` flag we
+will add the configuration for `crio` and `kubelet` to the final machine config
+manifest. Then the existing code path will apply the change as normal.
+
+#### API Server Admission Hook
+
+The existing admission hook has 4 checks when it comes to workload pinning.
+
+1. Check if `pod` is a static pod
+   - Skips modification attempt if it is static.
+1. Checks if currently running cluster topology is Single Node
+   - Skips modification if it is anything other than Single Node
+1. Checks if all running nodes are managed
+   - Skips modification if any of the nodes are not managed
+1. Checks what resource limits and requests are set on the pod
+   - Skips modification if QoS is guaranteed or both limits and requests are set
+   - Skips modification if after update the QoS is changed
+
+We will need to alter the code in the admission controller to remove the check
+for Single Node Topology, the other configurations should remain untouched.
+
+### Risks and Mitigations
+
+The sames risks and mitigations highlighted in [Management Workload
+Partitioning](management-workload-partitioning.md) apply to this enhancement as
+well.
+
+We need to make it very clear to customers that this feature is supported as a
+day 0 configuration and day n+1 alterations are not be supported with this
+enhancement. Part of that messaging should involve a clear indication that this
+should be a cluster wide feature.
+
+A risk we can run into is that a customer can apply a CPU set that is too small
+or out of bounds can cause problems such as extremely poor performance or start
+up errors. Mitigation of this scenario will be to provide proper guidance and
+guidelines for customers who enable this enhancement. Furthermore, the
+performance team would need to be reached out to for more specific information
+around upper and lower bounds of CPU sets for running an Openshift cluster.
+
+It is possible to build a cluster with the feature enabled and then add a node
+in a way that does not configure the workload partitions only for that node. We
+do not support this configuration as all nodes must have the feature turned on.
+However, there might be a race condition where a node is added and is in the
+process of being restarted with workload partitioning, during that process pods
+being admitted will trigger a warning. We expect the resulting error message
+described in [failure modes](#failure-modes) to explain the problem well enough
+for admins to recover.
```

Comment:
I would expect a node that is being rebooted to be cordoned first. Does that not happen?

Reply:
You're right that should still occur. This has more to do with _if_ the admission controller modifies the `resources` attribute of a pod or not. In the current logic we validate that all nodes are correctly advertising their Allocatable CPU resources with the appropriate `workload.openshift.io/cores`. If not we assume the feature has not been turned on for all nodes, or a node has failed to correctly be setup with partitioning.

An approach we could take here is to check if the node is missing the `Allocatable` field and if it is `Unschedulable` (cordoned off) and is updating, then we modify the pod spec as we would normally.

Reply:
This could be an expensive check if there are say 100 nodes in the cluster. I think we need to revisit if this is the check we want going forward.

Reply:
IIRC, when we talked about this as part of the other enhancement it was going to be a 1-way switch. If some global flag was set to say that partitioning should be used, then the API server would look at nodes each time a pod was admitted until it saw that all nodes had the feature enabled, then after that point it would always assume that partitioning was set up. 

I think the trouble we ran into with that approach was that it allowed some pods to be scheduled to a node without partitioning configured for those pods (for example, a daemon set or static pod that is scheduled early in the life of a node before all of the nodes in the cluster are reporting their special resource), and then later schedule other pods to the same node with partitioning configured. When that happens, you would end up in a state where some workloads would not be partitioned. This is what led to the requirement that the feature always be on from the beginning of the life of the cluster, and then we further simplified by saying only single-node would be supported.

So I agree with @browsell that we need to think about exactly what algorithm we're using to decide when the feature is "on" and how to ensure that partitioning is applied consistently and correctly to all pods.

Reply:
In regard to the global variable aspect of this, I'm thinking of a couple of approaches. 

1. We might want it to be a part of the [Infrastructure Status API](https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go#L51), like we currently do for topology. For the current single node clusters we can check if the node has the partitioning set, we can use the NTO to check and set that status.

Something like this, might be useful here.
```yaml
status:
    cpuPartitioning: ClusterPartitioned | UnPartitioned 
```

2. The other approach, is that we update the current Kubelet code to also add a label to the node that have partitioning configs, so we can use that as a signal since I think that will be set by startup time.

I think adding a label of `workload.openshift.io/managed: ''` to the node when it's partitioned might help us in solving the problem of timing. Applying a node affinity to incoming pods via the admission webhook when we detect it's on.